tracing-test/src/lib.rs

```diff
@@ -0,0 +1,87 @@
+//! Testing utilities for `tracing`
+//!
+//! # Overview
+//!
+//! `tracing-test` offers mock implementations of `tracing` types such as Spans, Events,
+//! and Subscribers. These mocks are intended to be used in testing `tracing`-based crates.
+//!
+//! # Usage
+//!
+//! Add the following to your `Cargo.toml`:
+//! ```toml
+//! tracing-test = "0.1"
+//! ```
+//!
+//! ## Mock subscriber
+//!
+//! ```rust
+//! # fn docs() {
+//! use tracing_test::*;
+//! use tracing::{debug, span, Level};
+//!
+//! let (subscriber, handle) = subscriber::expect()
+//!         .enter(span("foo"))
+//!         .event(
+//!             event().with_fields(field("message").with_value(&tracing::field::debug(
+//!                 format_args!("hello from my event! yak shaved = {:?}", true),
+//!             ))),
+//!         )
+//!         .exit(span("foo"))
+//!         .drop_span(span("foo"))
+//!         .done()
+//!         .run_with_handle();
+//!
+//! tracing::subscriber::with_default(subscriber, || {
+//!     let foo = span!(Level::DEBUG, "foo");
+//!     foo.in_scope(|| {
+//!         debug!("hello from my event! yak shaved = {:?}", true);
+//!     });
+//! });
+//! handle.assert_finished();
+//! # }
+//! ```
+#![cfg_attr(not(feature = "std"), no_std)]
```

Comment:
I don't know if it matters all that much to have `no-std` support for this crate...in most cases, I think tests are run with `std` even in projects that are otherwise `no-std`, with
```rust
#[cfg(test)]
extern crate std;
```
or similar.

On the other hand, I don't think we use any `std` APIs that aren't in `alloc`, so we _could_ support `no-std`. However, I think we should make sure that this actually builds without std, before publishing it with `no-std`. 

For example, I think that the import paths for the `std` APIs we use won't work, since they might be coming from `core`/`alloc`. If we want the test support code to support no-std, we need to do something like what `tracing` and `tracing-core` do: create [a module that conditionally re-exports from either `std` or `core`/`alloc`](https://github.com/tokio-rs/tracing/blob/master/tracing-core/src/stdlib.rs) and ensure it's used everywhere instead of `std`.